### PR TITLE
World Enhancement

### PR DESCRIPTION
**Pull request by Anatoly03**
This Pull Request will focus on making the user experience of the API better. Expectations/ Ideas:

---
- [ ] World Events?
- [ ] World Zones?
---
```js
const shiftGameLoop = (f => {
    // setup
    f.loop(g => {
        // round
    })
})
```

### CODE REVIEW

**Comment by Anatoly03** [comment]
- [x] Added edited source code for The Line Bot as an example.

**Comment by Anatoly03** [comment]
Added the following wrapper for commands:
```ts
client.onCommand('admin', p => p.isBotAdmin, callback)
client.onCommand('cmd', callback)
client.onCommand('help', () => 'Bot Help: !help !cmd !admin')
```

File: src/types/player.ts [hunk 1]
```diff
@@ -161,8 +159,28 @@ export class SelfPlayer extends Player {
...
+        if (preamble.length > MESSAGE_SIZE)
+            throw new Error('Chat preamble is larger than message size. Bad.')
+
+        // TODO regex?
+        if (content.length > CONTENT_ALLOWED_SIZE) {
+            const separator = content.substring(0, CONTENT_ALLOWED_SIZE).lastIndexOf(' ') || CONTENT_ALLOWED_SIZE
+            this.say(preamble, content.substring(0, separator))
+            this.say(preamble, content.substring(separator, content.length))
+            return
+        }
```
Comment (Anatoly03) [inline src/types/player.ts:195]: @ScorpioneOrzion This is what currently is happening, will rewrite to regex later

**Comment by Anatoly03** [comment]
The latest commits implemented the following:

```ts
export class StoredPlayer extends PlayerBase {
    static players: PlayerArray<StoredPlayer>
    constructor(args) { super(args) }

    public wins: number = 0
    public deaths: number = 0
    public rounds: number = 0

    static module(client: Client) {
        // Add commands here
        return client
    }
}
```
```
client.include(Modules.StorageModule('players.yaml', StoredPlayer))
```

is the base of creating the automatically up-to-date `StoredPlayer.players` which is a list of players that behaves as follows: Every player entry is of type `StoredPlayer` and accessing or setting the child members will be synchronised with `players.yaml`

**Comment by Anatoly03** [comment]
In the latest commit, the data structures now have a generic for mutability lock, as well as data being saved automatically on mutable methods.

```ts
    static module(client: Client) {

        client.on('player:join', ([p]) => {
            if (this.players.some(k => k.cuid == p.cuid)) return
            this.players.push(new StoredPlayer(p))
        })

        client.onCommand('wins', ([p]) => {
            const player = StoredPlayer.players.byCuid(p.cuid)
            if (player) return `Your medals: Gold ${player.gold}, Silver ${player.silver}, Bronze ${player.bronze}`
        })

        return client
    }
```

**Comment by Anatoly03** [comment]
Bug: Client no longer connects, will fix tomorrow.

**Comment by Anatoly03** [comment]
Ban Module, new bans now kicks player immediately.

**Comment by Anatoly03** [comment]
Added WormBot example

**Comment by Anatoly03** [comment]
The Storage system has been revamped and fixed.

```ts
export class StoredPlayer extends PlayerBase {
    public wins: number = 0
}

export class StoredPlayerManager extends PlayerStorage<StoredPlayer> {
    public override module(client: Client) {

        client.on('player:join', ([p]) => {
            if (this.byCuid(p.cuid) != undefined) return
            this.push(new StoredPlayer(p))
        })

        client.onCommand('wins', ([p]) => {
            const player = this.byCuid(p.cuid)
            if (player) return `Wins: ${player.wins}`
        })

        return client
    }
}
```

Then create an instance of `StoredPlayerManager` which will contain the array synchronized with the list. Include this instance as a client module

**Review by ScorpioneOrzion — APPROVED** [review]

**Comment by Anatoly03** [comment]
The example bots are now up to date

**Pull request merged**